android/src/main/java/com/geektime/rnonesignalandroid/RNOneSignal.java

```diff
@@ -53,64 +53,135 @@ of this software and associated documentation files (the "Software"), to deal
 
-    private HashMap<String, INotificationReceivedEvent> notificationReceivedEventCache;
-    private boolean hasSetNotificationWillShowInForegroundHandler = false;
+    private HashMap<String, INotificationWillDisplayEvent> notificationWillDisplayCache;
+    private HashMap<String, INotificationWillDisplayEvent> preventDefaultCache;
+
+    private boolean hasAddedNotificationForegroundListener = false;
+    private boolean hasAddedInAppMessageLifecycleListener = false;
+
+    private IInAppMessageClickListener rnInAppClickListener = new IInAppMessageClickListener() {
+        @Override
+        public void onClick(IInAppMessageClickEvent event) {
+            try {
+                IInAppMessageClickResult result = event.getResult();
+                sendEvent("OneSignal-inAppMessageClicked", RNUtils.convertHashMapToWritableMap(RNUtils.convertInAppMessageClickResultToMap(result)));
+            } catch (JSONException e) {
+                e.printStackTrace();
+            }
+        }
+    };
+
+    private IInAppMessageLifecycleListener rnInAppLifecycleListener = new IInAppMessageLifecycleListener() {
+        @Override
+        public void onWillDisplay(IInAppMessageWillDisplayEvent event) {
+            try {
+                sendEvent("OneSignal-inAppMessageWillDisplay", RNUtils.convertHashMapToWritableMap(RNUtils.convertInAppMessageToMap(event.getMessage())));
+            } catch (JSONException e) {
+                e.printStackTrace();
+            }
+        }
+
+        @Override
+        public void onDidDisplay(IInAppMessageDidDisplayEvent event) {
+            try {
+                sendEvent("OneSignal-inAppMessageDidDisplay", RNUtils.convertHashMapToWritableMap(RNUtils.convertInAppMessageToMap(event.getMessage())));
+            } catch (JSONException e) {
+                e.printStackTrace();
+            }
+        }
+
+        @Override
+        public void onWillDismiss(IInAppMessageWillDismissEvent event) {
+            try {
+                sendEvent("OneSignal-inAppMessageWillDismiss", RNUtils.convertHashMapToWritableMap(RNUtils.convertInAppMessageToMap(event.getMessage())));
+            } catch (JSONException e) {
+                e.printStackTrace();
+            }
+        }
+
+        @Override
+        public void onDidDismiss(IInAppMessageDidDismissEvent event) {
+            try {
+                sendEvent("OneSignal-inAppMessageDidDismiss", RNUtils.convertHashMapToWritableMap(RNUtils.convertInAppMessageToMap(event.getMessage())));
+            } catch (JSONException e) {
+                e.printStackTrace();
+            }
+        }
+    };
+
+    private INotificationClickListener rnNotificationClickListener = new INotificationClickListener() {
+        @Override
+        public void onClick(INotificationClickEvent event) {
+            try {
+                sendEvent("OneSignal-notificationClicked", RNUtils.convertHashMapToWritableMap(RNUtils.convertNotificationToMap(event.getNotification())));
```

Comment:
Pass the whole ClickEvent

Reply:
Addressed in [Update Android Notification and IAM listeners to send entire event](https://github.com/OneSignal/react-native-onesignal/pull/1532/commits/762ee2b05d3c4c6e272c7c4a3b994f0e00575b59)